add b-splines

ev-br opened the pull request:
This PR implements a base class for evaluation of b-splines without fitpack.
Representation is compatible with `splrep` and `splev` though.

The plan is to have a base class with a sensible API usable by both human users and interpolation / fitting to data (the latter is on an immediate plan).

matthew-brett commented:
Any progress here?  Is there anything I can do to help?

matthew-brett commented on scipy/interpolate/_bspl.pyx:
```diff
@@ -0,0 +1,410 @@
+        # nan
+        return -1
+
+    if (xval < tb) or (xval > te):
```
```
if (xval < tb) or (xval > te) and not extrapolate:
    return -1
``` 

?

matthew-brett commented on scipy/interpolate/_bspl.pyx:
```diff
@@ -0,0 +1,410 @@
+
+    # evaluate
+    interval = k
+    for ip in range(xp.shape[0]):
```
Can you `with nogil` this block?

matthew-brett commented on scipy/interpolate/_bspl.pyx:
```diff
@@ -0,0 +1,410 @@
+        B-spline coefficients
+    xp : ndarray, shape (s,)
+        Points to evaluate the spline at.
+    der : int
```
I guess this called `der` for compatibility with older code?  `dorder` clearer as name?

ev-br replied:
Renamed it `nu` to agree with *Poly and their subclasses (monotone interpolators).

matthew-brett commented on scipy/interpolate/_bspl.pyx:
```diff
@@ -0,0 +1,410 @@
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def _colloc(double[::1] x, double[::1] t, int k, double[::1, :] ab, int offset=0):
```
cdef function?

Can this be `nogil` ?

matthew-brett commented on scipy/interpolate/_bspl.pyx:
```diff
@@ -0,0 +1,410 @@
+            ab[kl + ku + j + offset - clmn, clmn] = wrk[a]
+
+
+def _handle_lhs_derivatives(double[::1]t, int k, double xval,
```
`cdef` ? `nogil` ?

matthew-brett commented on scipy/interpolate/_bspl.pyx:
```diff
@@ -0,0 +1,410 @@
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def _norm_eq_lsq(double[::1] x,
```
`cdef` / `cpdef` ?  `nogil` ?

matthew-brett commented on scipy/interpolate/_bspl.pyx:
```diff
@@ -0,0 +1,410 @@
+                
+            # ... and A.T @ y
+            for ci in range(rhs.shape[1]):
+                rhs[row, ci] = rhs[row, ci] + wrk[r] * y[j, ci] * wval    # XXX: y is not typed
```
Ah - y is not typed - hence this has to happen in Python space?

Is it possible speed benefits from enforcing type outweigh memory benefits?  Or can specialize for common types with fused type?

ev-br commented:
Thanks for the review @matthew-brett! I've sprinkled cython code with `with nogil` clauses. Nearly everything is done without the GIL now.

> Any progress here? Is there anything I can do to help?

I believe I addressed your review comments (GH does not hide some outdated diffs because of multiline statements).
Other than that, there was no activity since what, last November. From my POV, this is basically ready. There are quite a few enhancements possible, but the core functionality is there I think.
It would be great if you could kick its tires and see if this works for your use cases. 

Maybe it'd be good to merge (or decide we do not merge this) early in the release cycle, so that we have time to tweak things before the interface is frozen by backwards compat.

ev-br commented:
I'd also ping @insertinterestingnamehere who expressed some interest in splines at some point.

pv commented on scipy/interpolate/_bsplines.py:
```diff
@@ -0,0 +1,793 @@
+
+        """
+        if not self.t.flags.c_contiguous:
+            self.x = self.x.copy()
```
Should be `t` not `x`

pv commented on scipy/interpolate/_bsplines.py:
```diff
@@ -0,0 +1,793 @@
+        ct = len(self.t) - len(c)
+        if ct > 0:
+            c = np.r_[c, np.zeros((ct,) + c.shape[1:])]
+        tck = fitpack.splder((self.t, c, self.k), nu)
```
1D c only? Seems it could be easily addressed by adjusting splder/splantider code...

ev-br replied:
Indeed, done in https://github.com/ev-br/scipy/commit/446bf12e5458906cfb2bb27abc45708dd6f7e1b6 with some small cost in readability but hey.

pv commented:
This PR is doing it right, and contains the necessary basic functionality while preserving FITPACK compatibility. I don't see any major issues.

There are some interface quibbles that one can still ask:

- Shouldn't the `make_*` functions return BSpline and not tck?
- Having a tck attribute in BSpline that just returns t, c, k could be useful (which would mitigate the above if you really just want tck).
- Not sure I like the `make_*` names... Don't have better suggestions ATM though.

ev-br commented:
Regarding the interface:

* I don't much like the `make_*` names either, and I also don't have better ideas.
* Not sure what would be an advantage of having an additional tck attribute, given than t, c, and k attributes are public:

```
>>> b = BSpline.basis_element([1, 2, 3, 4])
>>> tck = (b.t, b.c, b.k)
```

* The motivation for `make_` functions to return tck rather than a BSpline object is extendability and ease of subclassing. For instance, a user can define a BSplineWithAsymptotics as a subclass (two-line override of _evaluate method), and use `make_interp_spline` with the correct values and derivatives at the boundaries.  OTOH, the alternative is not too taxing either, so the argument cuts both ways. I don't have a strong opinion here, and can change it if anybody has.

pv commented:
I think having two different spline object representation, tck-tuple and BSpline is not that great, even if they are essentially the same thing. I would prefer using BSpline everywhere in the interface. Note also that since there's no freestanding evaluation function, you'd usually just wrap the returned object into BSpline right away.

With `tck` attribute, reusability would be just:
```
bspl = make_interp_spline(...)
myspline = BSplineWithAsymptotics(*bspl.tck)
```
or just
```
myspline = BSplineWithAsymptotics.from_spline(bspl)
```

Since the data representation is compatible, it might be useful to also think whether many routines should accept/return BSpline objects instead of tck objects. For example. `splrep` and `splev` functions. Since they take tck as a single argument, it might be possible to retrofit them so that they also accept (and return?) BSpline objects. IOW, the spline-making functions in principle could be called `splrep` :) This probably has backward-compat issues, though, and those are always a bit nasty to work through.

ev-br commented:
The tck issue seems to circle back to your original idea of unpacking a BSpline object to a three-element tuple: 

```
b = make_interp_spline(x, y)
t, c, k = b

# or 
splev(b, x)
spalde(b, x)   # etc
```

With this, no change is needed for existing functions (splev, spder, splantider, spalde, sproot, splint).

Also `BSplineWithAsimptotics(b)` works without an extra `.tck` call.

This would be a bit clumsier: 
```
class NURBS(BSpline):
    ...   # use two first dims of the `c` array for numerator and denominator
```
but it's likely tolerable.

All in all, this is borderline too magical to my taste, but given backwards compat, alternatives seem worse:
* we can scrap the BSpline class, and keep from this PR the test suite (which, essentially, compares two independent implementations, one from FITPACK, one from __fitpack.h) and `make_` functions which return tck tuples.
* we can only add a `tck` tuple attribute and remove separate t, c, and k. This way lies `UnivariateSpline._data` :-)
* we can have *both* t, c and k as public attributes *and* a `tck` tuple getter. This is IMO already too much, and together with unpacking is definitely so.

As far as `sprep` et al are concerned: I don't think we want to pack it all to `splrep` and mess with backwards compat. 
I'd rather keep interpolation and fitting separate, and splrep does the latter by default.  This is unfortunate, but probably too late to change.

The fitting itself can come in several guises (LSQ vs (k+1)-order derivative jumps a la fitpack), we're likely to want some form of a local scheme at some point. FWIW, I think we're better off keeping factories separate for now, and maybe think about adding a wrapper when at least the majority of options are clearer.

ev-br commented:
> With this, no change is needed for existing functions (splev, spder, splantider, spalde, sproot, splint).

Scratch that.

For parametric curves, the meaning of trailing dimensions of `c` is transposed: 

```
In [2]: from scipy.interpolate import splrep, splev, BSpline

In [3]: x = np.arange(10)

In [4]: y = x**2

In [5]: t, c, k = splrep(x, y)

In [6]: splev(5., (t, c, k))
Out[6]: array(24.99999999999999)

In [7]: BSpline(t, c, k)(5.)
Out[7]: array(24.99999999999999)

In [8]: c1 = np.c_[c, c, c]

In [9]: splev(5., (t, c1, k))
Out[9]: 
[array(0.3333333333333328),
 array(3.1111111111111116),
 array(9.833333333333336),
 array(17.166666666666664),
 array(26.499999999999986),
 array(38.55555555555556),
 array(55.66666666666665),
 array(71.00000000000003),
 array(67.5),
 array(0.0),
 array(0.0),
 array(0.0),
 array(2.9e-322),
 array(2.37716983e-316)]

In [10]: splev(5., (t, c1[0], k))
Out[10]: array(0.3333333333333328)

In [11]: BSpline(t, c1, k)(5.)
Out[11]: array([ 25.,  25.,  25.])

In [12]: splev(5., (t, c1.T, k))
Out[12]: [array(24.99999999999999), array(24.99999999999999), array(24.99999999999999)]
```

ev-br commented:
Pushed the updated version, where

* `spl`* functions accept and return BSpline instances
* Users who start with the data points and only do `spl = splrep(x, y)` followed by `splev(xnew, spl)` will not notice anything.
* BSpline instances behave as tuples of (t, c, k) w.r.t. unpacking and indexing. I'm not a fan, but this is required for backwards compat

Demo:

```
In [3]: x = np.linspace(0, 10, 11)

In [4]: y = x**2

In [5]: spl = splrep(x, y)

In [6]: type(spl)
Out[6]: scipy.interpolate._bsplines.BSpline

In [7]: splev(x, spl)
Out[7]: 
array([  3.44265545e-17,   1.00000000e+00,   4.00000000e+00,
         9.00000000e+00,   1.60000000e+01,   2.50000000e+01,
         3.60000000e+01,   4.90000000e+01,   6.40000000e+01,
         8.10000000e+01,   1.00000000e+02])

In [8]: spl(x)
Out[8]: 
array([  3.44265545e-17,   1.00000000e+00,   4.00000000e+00,
         9.00000000e+00,   1.60000000e+01,   2.50000000e+01,
         3.60000000e+01,   4.90000000e+01,   6.40000000e+01,
         8.10000000e+01,   1.00000000e+02])

In [9]: t, c, k = spl

In [10]: t
Out[10]: 
array([  0.,   0.,   0.,   0.,   2.,   3.,   4.,   5.,   6.,   7.,   8.,
        10.,  10.,  10.,  10.])
```

* tck-tuples are also accepted by all functions in a backwards-compatible way
* the functions which both accept and return splines (splder, splantider and insert) preserve the type of intput: if you call them with a tck-tuple, you get a tck-tuple back. 

A potentially sticky point is coefficient arrays with D>1: Fitpack interpolates along the last axis, while BSpline interpolates along the *first* axis --- the latter is consistent with PPoly. So I do not see a way of making BSpline consistent with both.

If you get your spline curve from `splprep` and then `splev` it or `__call__` it, you're in the clear.
To run into issues, a user needs to construct a BSpline, unpack it unto  a tuple, and then try to `splev` the tuple. In this case, the coefficient array has to be transposed. This is hopefully a rare corner case. Thoughts?

ev-br commented:
Two more changes in the last installment:
* `BSpline.tck` readonly attribute, as requested.
* `make_` functions return BSplines. 

There's still some overlap between `splrep`/`splprep` and `make_interp_spline`/`make_lsq_spline`; I still don't see  a clean way of merging these two together, and I still don't think they should be combined. @pv?

matthew-brett commented:
Where are we with this one?   What are the blockers to merging?

ev-br commented:
Rebased.

@matthew-brett I think I've addressed most of @pv's concerns starting from https://github.com/scipy/scipy/pull/3174#issuecomment-120959918.
Code review for commits  starting from  469c8f7 would be great. They deal with the interface, esp on the interaction of of tck- and BSpline representations, and the overlap between splrep/splprep and new make_* functions. 

This can potentially be somewhat disruptive downstream since it touches both spl* interface and interp1d. FWIW I don't expect serious backwards compat regressions. There are two ways downstream code can get into problems:
* if you rely on specific numeric results produced by interp1d. I don't believe we guarantee specific numbers in a way `numpy.random` does, but if someone hardcoded the numbers, they'll change. If that turns out to be a problem, we can revert back to using splmake.
* if you do some manual manipulations stacking coefficients produced by splrep et al. This is likely a corner case, and I did not find any examples in the wild. If there are issues, downstream will need to use workarounds depending on the scipy version.

All in all, I think this is ready and I'd take a plunge :-).

ev-br commented:
closed and reopened to start tests

ev-br commented:
@matthew-brett In case I wasn't clear (at 2am local time): I think this PR is ready. New features are out of scope, review comments are addressed. I'm not going to self-merge it though. Further activity here is going to be due to further review, if any.

pv commented:
I have some minor comments (not using `assert_equal` for fp comparisons, using `BSpline._construct_fast` instead of `BSpline` in a number of places).

The remaining bigger issue is probably with the transposition in the FITPACK interface. AFAICS, this should be understood as something specific to `splprep`. The fitpack_impl code also handles the 2D case, and also this is done via `return list(map(lambda c, x=x, t=t, k=k, der=der: splev(x, [t, c, k], der, ext), c))`, not too efficient, and returns list instead of array...

I think it is not very pretty that `splev(x, bspl)` is not equivalent to `bspl(x)`. I think it's likely people will write `splev(x, bspl)` and be surprised when it returns unexpected results, works slowly, or just fails to work (e.g. ndim > 3).

One way around this would be to make `splprep` return a special subclass
`class PBSpline(BSpline): pass` that just transposes its output axis (and this behavior inherited by `splev`). Then `splev(x, bspl)` can be understood to be equivalent to `bspl(x)`.

The remaining question in that case is then on what to do with the more rich extrapolation options in `splev` --- maybe support only for 1D for now?

ev-br commented:
I'm not sure I understand your suggestion. Would you please elaborate @pv?

ev-br commented:
OK, the idea of a dummy subclass to serve as a marker, unless I miss something simple, takes you quite far, but still cannot fix the following: given `b, u = splprep(x)`, you can either have `splev(u, b) == b(u)` or `splev(u, b) == x`, but not both. Here's a branch which implements this: https://github.com/ev-br/scipy/commits/bsplines_ng_2 (the relevant commit is https://github.com/ev-br/scipy/commit/bce7c9b274e3894714d4572ea634f39c9685ad96)

However, if a `BSpline` class has an `axis` attribute to specify the interpolation axis, then `splprep` simply creates an objects with `axis=1` and all seems well. This is what this PR now does, module a couple of small wrinkles I'm going to iron over (some more tests; decide what splev et al do when given a BSpline with axis != 0 etc). Meanwhile a review of of e4aef54 and da56a3d would be really appreciated.

ev-br commented:
I've ticked all the remaining TODOs, I think all review comments are addressed, the friction at the FITPACK interface is at a minimum. I'm done for now, this PR's status is *needs review*.

matmel commented on scipy/interpolate/_bsplines.py:
```diff
@@ -0,0 +1,879 @@
+
+    if x.ndim != 1 or np.any(x[1:] - x[:-1] <= 0):
+        raise ValueError("Expect x to be a 1-D sorted array_like.")
+    if x.shape[0] < k+1:
```
Hello,

This is the first time I make a pull request comment, so I accept any guidance on how to make this more effectively.

This test:
```python
if x.shape[0] < k+1:
    raise ValueError("Need more x points.")
```
is not necessary (I think) and prevent to use the function `make_interp_spline` in legitimate cases. For example if you interpolate 2 points with a cubic spline and that you impose left and right first derivatives, your spline is fully defined. (2 points + 2 derivatives = 4 conditions for the 4 control points required by a cubic spline). This use case is commonplace when you need to interpolate splines through points for geometric constructions (ala CAD software).

Note also that AFAICT if you do not define enough conditions (i.e. number of points and derivatives), the test L.686 will catch it:
```python
if nt - n != nleft + nright:
    raise ValueError("number of derivatives at boundaries.")
```
Finally, you can prove my point with the following test (with the incriminated test commented out):
```python
import numpy as np
from numpy.testing import assert_allclose
from scipy.interpolate import make_inter_spline

# interpolation of f(x) = x**3 => f'(0) = 0, f'(1) = 3
x = [0., 1.]
y = [0., 1.]
deriv_l = [(1, [0.])]
deriv_r = [(1, [3.])]
s = make_interp_spline(x, y, k=3, deriv_l=deriv_l, deriv_r=deriv_r)

xx = np.linspace(0., 1,.)
assert_allclose(xx**3, s(xx), rtol=1e-14)
```

Thanks

matmel commented:
Hello, I commented the final diff to ask for removing an unnecessary (I think) test that prevent legitimate use cases. 

I send this comment because the code comment itself does not seem to send an e-mail to watchers of this PR.

I'm new to github comments so sorry in advance if it generates too much noise...

ev-br commented:
Thank you @matmel! I agree that restriction is extraneous and could be removed.
Would you be interested in sending a PR with the fix? 
The example you gave should serve as a test case, please add it to `test_bsplines.py`

Normally, all PRs should be against master, but since this is itself still in a PR, it'd be ideal if you could send a PR against my scipy fork, https://github.com/ev-br/scipy/tree/bsplines, or put a relevant commit to your scipy fork where I can cherry-pick it from.
This way, you authorship is going to be preserved :-). 

Regarding the format, it's very welcome to comment on the open PRs and github does send email notifications on such comments (I received it, at least).

Any further review would be most welcome, too.

ev-br commented:
The CI failure is baffling. I cannot reproduce this ImportError locally using a range of python/numpy versions, with or without wheels: locally everything imports just fine and all tests pass.

The only suspicion I have is that something went south with TravisCI caching, but I've no idea how to temporarily disable it or work around it.

rgommers commented:
I see the issue locally, will send you a PR for build changes.

rgommers commented:
There's some test noise:

    too many axes: 2 (effrank=2), expected rank=1
    too many axes: 2 (effrank=2), expected rank=1

and one test failure (on 32-bit Linux):

    ======================================================================
    FAIL: test_spalde (test_bsplines.TestInteropSplPrep)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/rgommers/Code/bldscipy/build/testenv/lib/python2.7/site-packages/scipy/interpolate/tests/test_bsplines.py", line 795, in test_spalde
        assert_allclose(spalde(0.1, self.b), all_der, atol=1e-15)
      File "/home/rgommers/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 1391, in assert_allclose
        verbose=verbose, header=header)
      File "/home/rgommers/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 733, in assert_array_compare
        raise AssertionError(msg)
    AssertionError: 
    Not equal to tolerance rtol=1e-07, atol=1e-15

    (mismatch 16.6666666667%)
     x: array([[  6.283185e-01,   5.877853e-01,   8.090170e-01],
           [  6.283185e+00,   5.083701e+00,  -3.690573e+00],
           [  4.440892e-13,  -2.345007e+01,  -3.236774e+01],
           [  1.056932e-11,  -2.270116e+02,   9.741356e+01]])
     y: array([[  6.283185e-01,   5.877853e-01,   8.090170e-01],
           [  6.283185e+00,   5.083701e+00,  -3.690573e+00],
           [  4.618528e-13,  -2.345007e+01,  -3.236774e+01],
           [  1.101341e-11,  -2.270116e+02,   9.741356e+01]])

rgommers commented:
Hmm, I don't see your fork Evgeni. IIRC I had that problem before. So cherry-picking it is, here's the build fix: https://github.com/rgommers/scipy/commit/3fed1469ddfb97ae213943a014ff5cf33c5b753e

ev-br commented:
Thanks Ralf! I cherry-picked your build fix. That's real deep distutils magic, it looks so natural after the fact. Thanks!

rgommers commented:
> That's real deep distutils magic, it looks so natural after the fact.

That's nowhere near the real black magic:) Stuff like gh-7235 is where it gets fun.....

ev-br commented:
I've also bumped the test tolerance, the test should now pass on 32 bit as well. "Too many axes" noise is sadly not so easy to get rid of. It comes from this test:
https://github.com/scipy/scipy/pull/3174/files#diff-510f85187efa22ea6199e8ac5a71e1e3R610

        # test that both "old" and "new" splrep raise for an n-D ``y`` array
        # with n > 1
        y2 = np.c_[y, y]
        assert_raises(Exception, splrep, x, y2)
        assert_raises(Exception, _impl.splrep, x, y2)

Both a non-very-descriptive exception and a print come from deep inside the Fortran wrappers, similar to the  "0-th dimension must be fixed to 4 but got 2" noise which comes from testing BLAS wrappers. 
I can take the test off, do you think it's preferrable?

rgommers commented:
Test indeed passes now. I see where that noise comes from now. I think disabling tests isn't the right thing to do; instead we should probably fix ``f2py`` to give proper Python warnings instead of using ``fprintf``. Then we can fix the BLAS wrapper noise as well.

pv commented:
Didn't look where the too many axes comes here, but maybe you need to `ravel()`?

rgommers commented:
It's from 

    ./numpy/f2py/src/fortranobject.c:
      922 :                 fprintf(stderr,"too many axes: %d (effrank=%d), expected rank=%d\n",

as is the BLAS stuff:

      860 :                     fprintf(stderr,"%d-th dimension must be fixed to %" NPY_INTP_FMT

ev-br commented:
> maybe you need to ravel()?

I don't think so. The purpose of the test is to assert that `splev` (that is, the original splev) does not accept 2D arrays, and that this PR does not change this:

```
(Pdb) y2.shape
(41, 2)
(Pdb) splrep(x, y2)
too many axes: 2 (effrank=2), expected rank=1
*** error: failed in converting 3rd argument `y' of dfitpack.curfit to C/Fortran array
```

Whether documenting the status quo like this is useful --- I think it is, but not enough to argue for it if there are strong preferences to the contrary :-).

ev-br commented:
Rebased. 

Also made the interface for the boundary conditions in `make_interp_spline` consistent with CubicSpline.

pv commented on scipy/interpolate/fitpack.py:
```diff
@@ -651,21 +413,15 @@ def splint(a, b, tck, full_output=0):
+    if isinstance(tck, BSpline):
+        t, c, k = tck
+
+        # FITPACK expects the interpolation axis to be last, so roll it over
```
This comment is a bit inaccurate --- FITPACK doesn't do the interpolation, instead it's done on Python side via `list(map(lambda c, t=t, k=k, mest=mest: sproot([t, c, k], mest), c))`, which is not really an optimal way to do it.

Again, I'd suggest deprecation warning for the multidim case, and a message suggesting `spl.integrate` instead.

pv commented on scipy/interpolate/_bsplines.py:
```diff
@@ -0,0 +1,885 @@
+        ct = len(self.t) - len(c)
+        if ct > 0:
+            c = np.r_[c, np.zeros((ct,) + c.shape[1:])]
+        t, c, k = _fitpack_impl.splantider((self.t, c, self.k), 1)
```
Maybe we should call `splint` if the spline is 1D (and extrapolate=False since splint assumes zero extension)? May be faster?

pv commented on scipy/interpolate/fitpack.py:
```diff
@@ -559,68 +331,54 @@ def splev(x, tck, der=0, ext=0):
+    if isinstance(tck, BSpline):
+        t, c, k = tck
+
+        # fitpack interpolates along the last axis
```
I would actually suggest a DeprecationWarning here --- the user should call `bspline(x)` instead of getting the sub-optimal `list(map(... ))` Python-side workaround.

The sub-optimal behavior is OK if tuples passed in, however.

pv commented on scipy/interpolate/fitpack.py:
```diff
@@ -213,106 +144,18 @@ def splprep(x, w=None, u=None, ub=None, ue=None, k=3, task=0, s=None, t=None,
-            return tcku, fp, ier, _iermess['unknown'][0]
+        tcku, fp, ier, mesg = res
+        (t, c, k), u = tcku
+        spl = BSpline(t, c, k, axis=1)
```
I'd like to chicken out on returning BSpline objects instead of tuples by default from `splprep`, `splrep` since it is a backward compat break. Or, put it behind a keyword argument that we can turn ON in some future major release.

Supporting BSpline in `splev`, `splint` et al. however is ok.

pv commented on scipy/interpolate/fitpack.py:
```diff
@@ -433,95 +277,22 @@ def splrep(x, y, w=None, xb=None, xe=None, k=3, task=0, s=None, t=None,
-        except KeyError:
-            return tck, fp, ier, _iermess['unknown'][0]
+        tck, fp, ier, mesg = res
+        spl = BSpline(*tck)
```
ditto re chickening out

pv commented on scipy/interpolate/fitpack.py:
```diff
@@ -690,47 +447,38 @@ def sproot(tck, mest=10):
+        # NB: This transpose is a no-op if c is 1D.
+        sh = tuple(range(c.ndim))
+        c = c.transpose(sh[1:] + (0,))
+        return _impl.sproot((t, c, k), mest)
```
As above

pv commented on scipy/interpolate/__init__.py:
```diff
@@ -87,6 +87,15 @@
    splantider
    insert
 
+An alternative set of univariate spline routines:
```
I'd put these together with the splrep list above since the objects are compatible.
Instead, the UnivariateSpline stuff should be presented as the "Alternative set".

ev-br commented:
OK, pushed an update where I believe I addressed @pv's review comments. Apparently Github does not hide outdated comments anymore. Relevant commits start from  https://github.com/scipy/scipy/pull/3174/commits/e6c5b7272472b1e1810f10c8b89a4c9350e38190.

The main change is that `splrep` and `splprep` again return tck tuples and not BSpline objects (https://github.com/scipy/scipy/pull/3174/commits/0131d72f0f7a5c13f00961aea054094b2b78f071). I've to say I actually like it more. The interface with `spl*`-functions is still not perfect for multidimentional splines, but at least the logic seems clear: `splint(BSpline, ...)` is consistent with `BSpline.integrate()`, and `splint(tck, ...)` is backwards compatible. So far I did not implement a keyword switch, as suggested  in https://github.com/scipy/scipy/pull/3174#discussion_r79290152, but it's easy to bolt on. 

Also, tuple-like unpacking of BSplines is gone. This was needed for backwards compat if `splrep` returns a BSpline instance, to support legacy code which does `t, c, k = splrep(x, y)`. But as long as we're not trying to sneak BSplines in place of tuples, I'd rather avoid that sort of magic. Knots, coefficients and the spline order are anyway available as `b.tck` or as individual attributes `b.t`, `b.c` `b.k`.

pv commented:
Ok, looks good to me currently, and OK to merge from my side.
If we come up with better names than `make_*` or other tweaks before 0.19.0, we can make those.

ev-br commented:
OK, I'm going to hit the green button in a couple of days, unless further comments.

matthew-brett commented:
Great - thanks very much for your persistence on this one.

pv commented:
Thanks @ev-br, let's get it in!

ev-br commented:
Thanks all for the code, for the reviews, for encouragement and for evolving this from where it started to a useful state!

rgommers commented:
Yay! Great job & persistence @ev-br. 

Maybe a PR to say "this is the way of the future" and clarify some of the existing text on splines in the roadmap would be a good idea?

ev-br commented:
Here's a PR: https://github.com/scipy/scipy/pull/6715

pv commented:
Note that this introduced some performance regressions:
https://pv.github.io/scipy-bench/regressions.xml
e.g. https://pv.github.io/scipy-bench/index.html#interpolate.Interpolate1d.time_interpolate?commits=a247ff7f61e6d88a02db3f451c52b378a3238c29&idx=3

pv commented:
Or more precisely, the `slinear` and `zero` got slower --- the cubic and quadratic are faster.
Maybe there's some trivial special casing possible.

ev-br commented:
Hmm, evaluations for `slinear` and `zero` got about an order of magnitude faster, (interp1d_eval benchmark), but construction (time_interpolate benchmark) got an order of magnitude slower, or am I misreading it?

pv commented:
Note that there are some performance issues that may need looking into:
https://pv.github.io/scipy-bench/regressions.xml
